pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaTableSource.scala

```diff
@@ -0,0 +1,164 @@
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import java.util.function.Consumer
+import akka.stream.stage.{AsyncCallback, GraphStageLogic, GraphStageWithMaterializedValue, OutHandler, StageLogging}
+import akka.stream.{Attributes, Outlet, SourceShape}
+import akka.Done
+import akka.annotation.InternalApi
+import akka.event.Logging
+
+import scala.concurrent.{Future, Promise}
+import scala.util.control.NonFatal
+import akka.stream.ActorAttributes
+import akka.stream.alpakka.pravega.TableSettings
+import io.pravega.client.KeyValueTableFactory
+import io.pravega.client.tables.{IteratorItem, KeyValueTable, KeyValueTableClientConfiguration, TableEntry}
+
+import scala.collection.mutable
+
+import java.util.concurrent.Semaphore
+import io.pravega.common.util.AsyncIterator
+@InternalApi private final class PravegaTableSourceStageLogic[K, V](
+    shape: SourceShape[(K, V)],
+    val scope: String,
+    tableName: String,
+    keyFamily: String,
+    tableSettings: TableSettings[K, V],
+    startupPromise: Promise[Done]
+) extends GraphStageLogic(shape)
+    with StageLogging {
+
+  override protected def logSource = classOf[PravegaTableSourceStageLogic[K, V]]
+
+  private def out = shape.out
+
+  private var keyValueTableFactory: KeyValueTableFactory = _
+  private var table: KeyValueTable[K, V] = _
+
+  private val queue = mutable.Queue.empty[(K, V)]
+
+  private val semaphore = new Semaphore(tableSettings.maximumInflightMessages)
+
+  private var closing = false
+
+  val logThat: AsyncCallback[String] = getAsyncCallback { message =>
+    log.info(message)
+  }
+
+  private def pushElement(out: Outlet[(K, V)], element: (K, V)) = {
+    push(out, element)
+    semaphore.release()
+  }
+
+  val onElement: AsyncCallback[(K, V)] = getAsyncCallback[(K, V)] { element =>
+    if (isAvailable(out) && queue.isEmpty)
+      pushElement(out, element)
+    else
+      queue.enqueue(element)
+
+  }
+
+  val onFinish: AsyncCallback[Unit] = getAsyncCallback[Unit] { _ =>
+    closing = true
+    if (queue.isEmpty)
+      completeStage()
+
+  }
+
+  setHandler(
+    out,
+    new OutHandler {
+      override def onPull(): Unit = {
+        if (!queue.isEmpty)
+          pushElement(out, queue.dequeue())
+        if (closing && queue.isEmpty)
+          completeStage()
+      }
+    }
+  )
+
+  def nextIteration(iterator: AsyncIterator[IteratorItem[TableEntry[K, V]]]): Unit =
+    iterator.getNext
+      .thenAccept(new Consumer[IteratorItem[TableEntry[K, V]]] {
+        override def accept(iteratorItem: IteratorItem[TableEntry[K, V]]): Unit = {
+          if (iteratorItem == null) {
+            onFinish.invoke(())
+          } else {
+            iteratorItem.getItems.stream().forEach { tableEntry =>
+              semaphore.acquire()
+              onElement.invoke((tableEntry.getKey.getKey, tableEntry.getValue))
+            }
+            nextIteration(iterator)
+          }
+        }
+      })
```

Comment:
Any reason why streams `AsyncCallbacks` aren't used for this code?

Reply:
I thought I was doing the right thing L89 and L93 with AsyncCallbacks :(